File: 2021/parallel_and_other_perf_demos/mpi/python/02_integral.py
from mpi4py import MPI
import os, sys
from math import cos, pi

host = os.uname()[1]

comm = MPI.COMM_WORLD
my_rank = comm.Get_rank()
p = comm.Get_size()

def f(x):
    return cos(x)

def integrate(a, n, h):
    print("Starting from %f and doing %d steps with step %f" % (a, n, h))
    res = 0
    h2 = h/2

    for i in range(0, n):
        x = a + i * h
        res += f(x + h2) * h

    return res

a = 0.0
b = pi / 2
n = 100000
if len(sys.argv) > 1:
    n = int(sys.argv[1])

print("We are going to do %d steps" % n)

h = (b-a)/n
local_n = n/p

local_a = a + my_rank * local_n * h
local_b = local_a + local_n * h

integral = integrate(local_a, int(local_n), h)


if my_rank == 0:
    total = integral
    for s in range(1, p):
        integral = comm.recv(source=s)
        total = total + integral
else:
    comm.send(integral, dest = 0)

if my_rank == 0:
    print("res = %f" % total)

MPI.Finalize

File: 2022/parallel_and_other_perf_demos/mpi/cpp/02_integral.cpp
#include <mpi.h>
#include <math.h>
#include <stdio.h>

float fct(float x)
{
      return cos(x);
}

float integral(float a, long long n, float h);

int main(int argc, char *argv[])
{
      long long n, i, j, ierr,num;
      float h, result, a, b, pi;
      float my_a, my_range;

      int myid, source, dest, tag, p;
      MPI_Status status;
      float my_result;

      pi = acos(-1.0);  /* = 3.14159... */
      a = 0.;           /* lower limit of integration */
      b = pi*1./2.;     /* upper limit of integration */
      n = 100000;          /* default number of increment within each process */
      if(argc > 1) {
            n = atoi(argv[1]);
      }
      printf("Doing %d steps\n", n);

      dest = 0;         /* define the process that computes the final result */
      tag = 123;        /* set the tag to identify this particular job */

      /* Starts MPI processes ... */

      MPI_Init(&argc,&argv);              /* starts MPI */
      MPI_Comm_rank(MPI_COMM_WORLD, &myid);  /* get current process id */
      MPI_Comm_size(MPI_COMM_WORLD, &p);     /* get number of processes */

      h = (b-a)/n;    /* length of increment */
      num = n/p;	/* number of intervals calculated by each process*/
      my_range = (b-a)/p;
      my_a = a + myid*my_range;
      my_result = integral(my_a,num,h);

      printf("Process %d has the partial result of %f\n", myid,my_result);

      if(myid == 0) {
        result = my_result;
        for (i=1;i<p;i++) {
          source = i;           /* MPI process number range is [0,p-1] */
          MPI_Recv(&my_result, 1, MPI_REAL, source, tag,
                        MPI_COMM_WORLD, &status);
          result += my_result;
        }
        printf("The result = %f\n",result);
      }
      else
        MPI_Send(&my_result, 1, MPI_REAL, dest, tag,
                      MPI_COMM_WORLD);      /* send my_result to intended dest.
                      */
      MPI_Finalize();                       /* let MPI finish up ... */

      return 0;
}


float integral(float a, long long n, float h)
{
      printf("Starting from %f and doing %d steps with step %f\n", a, n, h);
      long long j;
      float h2, aij, integ;

      integ = 0.0;                 /* initialize integral */
      h2 = h/2.;
      for (j=0;j<n;j++) {             /* sum over all "j" integrals */
        aij = a + j*h;        /* lower limit of "j" integral */
        integ += fct(aij+h2)*h;
      }
      return (integ);
}

